go/mysqlconn/client.go

```diff
@@ -0,0 +1,455 @@
+package mysqlconn
+
+import (
+	"crypto/sha1"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/youtube/vitess/go/sqldb"
+)
+
+// Connect creates a connection to a server.
+// It then handles the initial handshake.
+//
+// If context is canceled before the end of the process, this function
+// will return nil, ctx.Err().
+//
+// FIXME(alainjobart) once we have more of a server side, add test cases
+// to cover all failure scenarios.
+func Connect(ctx context.Context, params *sqldb.ConnParams) (*Conn, error) {
+	netProto := "tcp"
+	addr := ""
+	if params.UnixSocket != "" {
+		netProto = "unix"
+		addr = params.UnixSocket
+	} else {
+		addr = net.JoinHostPort(params.Host, fmt.Sprintf("%v", params.Port))
+	}
+
+	// Figure out the character set we want.
+	characterSet, err := parseCharacterSet(params.Charset)
+	if err != nil {
+		return nil, err
+	}
+
+	// Start a background connection routine.  It first
+	// establishes a network connection, saves it into 'c', then
+	// works on the handshake. 'mu' protects 'c', so if the
+	// context is canceled we can safely read 'c'.
+	done := make(chan error, 1)
+	mu := sync.Mutex{}
+	var c *Conn
+	go func() {
```

Comment:
It may be cleaner to have this goroutine not check the context, and just push a `struct{c, err}` on a result channel.
Then the outer check will return `c/err` if `result` fires. If `Done` fires, it starts a goroutine that waits on `result` and close c if it was eventually opened.